toast-tk-web-agent/src/main/java/io/toast/tk/agent/web/AgentServerImpl.java

```diff
@@ -12,46 +15,54 @@
 
 import io.toast.tk.agent.ui.IAgentApp;
 import io.toast.tk.core.agent.interpret.WebEventRecord;
+import io.toast.tk.core.rest.HttpRequest;
 import io.toast.tk.core.rest.RestUtils;
+import org.asynchttpclient.ws.WebSocket;
+import org.asynchttpclient.ws.WebSocketTextListener;
+import org.asynchttpclient.ws.WebSocketUpgradeHandler;
 
-public class AgentServerImpl  implements IAgentServer{
+public class AgentServerImpl implements IAgentServer {
 
-	private static final Logger LOG = LogManager.getLogger(RestRecorderService.class);
-	private String hostName;
+	private static final Logger LOG = LogManager.getLogger(AgentServerImpl.class);
+	private final ToastAsyncHttpClientProvider wsSlientProvider;
 	private IAgentApp app;
-	
+
 	@Inject
-	public AgentServerImpl(IAgentApp app){
-		try {
-			this.app = app;
-			this.hostName = InetAddress.getLocalHost().getHostName();
-		} catch (UnknownHostException e) {
-			LOG.error(e.getMessage(), e);
-		}
+	public AgentServerImpl(IAgentApp app, ToastAsyncHttpClientProvider wsSlientProvider){
+		this.app = app;
+		this.wsSlientProvider = wsSlientProvider;
 	}
 
-	public void sendEvent(WebEventRecord eventRecord, String ApiKey) {
+	public void sendEvent(WebEventRecord eventRecord, String apiKey) {
 		String json = new Gson().toJson(eventRecord);
-		RestUtils.postWebEventRecord(getWebAppURI()+"/api/record", json, ApiKey);
+		String url = getWebAppURI() + "/api/record";
+		RestUtils.postWebEventRecord(buildRequest(url, json, apiKey));
 	}
 
-	public boolean register() {
-		return register(null);
-	}
-	public boolean register(String ApiKey) {
+	public boolean register(String apiKey) {
 		try {
 			String url = getWebAppURI() + "/api/susbcribe";
 
+			// request.getHeader("X-FORWARDED-FOR")
 			String localAddress = Inet4Address.getLocalHost().getHostAddress();
-			AgentInformation info = new AgentInformation(localAddress, ApiKey);
+			AgentInformation info = new AgentInformation(localAddress, apiKey);
 			String json = new Gson().toJson(info);
-			
-			boolean isRegistered = RestUtils.registerAgent(url, json, ApiKey);
-			
+
+			LOG.info("Registering driver: {}", json);
+
+			boolean isRegistered = RestUtils.registerAgent(buildRequest(url, json, apiKey));
+
 			if(isRegistered) {
-				LOG.info("Agent registred with hotname {}", hostName);
-			}
-			else {
+				LOG.info("Agent registred with hotname {}", localAddress);
+				//Open Alive WebSocket
+				try {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Extract this nested try block into a separate method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS1141)